datapath: always generate BTF debug information

jibi opened the pull request:
See commits description.

brb commented on install/kubernetes/cilium/templates/cilium-configmap.yaml:
```diff
@@ -217,6 +217,10 @@ data:
 
+{{- if .Values.bpf.compileDebug }}
+  # Enable debugging of the BPF compilation process.
+  bpf-compile-debug: "true"
```
I'm going to impersonate @aanm : I know this is not related to your PR, but can't we move this flag to `--debug-verbose` which would reduce cilium-agent flags count by one?

jibi replied:
I would keep these options separated (at least for now) as:
* my understanding is that they provide debug visibility into different areas of the datapath logic
* `--bpf-compile-debug` may not work on a setup that does not use the Cilium docker images (because of potentially unsupported versions of iproute2/llvm)

so I can see how a user may want to enable `--debug-verbose=datapath` without having the debug symbols in their BPF program

aanm replied:
I'm going to impersonate myself:
We could have `--debug-verbose=datapath` and `--debug-verbose=datapath-compile` for without and with debug symbols in their BPF program.

jibi replied:
Right, sorry, I assumed you wanted to move this into `--debug-verbose=datapath`

pchaigno replied:
Is there any reason we shouldn't enable this by default? I doubt anyone is running Cilium with their own version of iproute2 or LLVM (the first because our fork is needed, the second because it would generate too many verifier issues).

jibi replied:
> I doubt anyone is running Cilium with their own version of iproute2 or LLVM (the first because our fork is needed, the second because it would generate too many verifier issues).

Even if that's the case for the end users (I wasn't aware of that but it's good to know) that may not be true in the context of development/testing. Are we fine also in that context to have a single option which controls both debug features of the datapath as well as generation of debug information? (Genuine question as I'm not opposed to that)

pchaigno replied:
Just to clarify: I was not suggesting a single option but either 1) a lack of option and `-g` always enabled or 2) an option to opt-out from `-g`.

jibi replied:
OK now I understand :+1:
I would avoid option 1) for the the concerns I expressed in the previous comment.
For option 2) I think that could work but it's probably not going to fit into `--debug-verbose` (so for this we would need to ~add another option~ replace `--bpf-compile-debug` with something else).

@aanm @brb I'd like to hear your thoughts as well

aanm commented on daemon/cmd/daemon_main.go:
```diff
@@ -228,9 +229,6 @@ func init() {
 	option.BindEnv(option.CGroupRoot)
 
-	flags.Bool(option.BPFCompileDebugName, false, "Enable debugging of the BPF compilation process")
-	option.BindEnv(option.BPFCompileDebugName)
```
This is a released flag, we should mark it for removal in the next Cilium release 1.11.
Right now, all we can do we is to mark this flag as deprecated and leave it here.
This also needs documentation updates in the upgrade guide.

aanm commented on pkg/datapath/loader/compile.go:
```diff
@@ -379,7 +377,8 @@ func compileDatapath(ctx context.Context, dirs *directoryInfo, isHost, debug boo
 // Compile compiles a BPF program generating an object file.
 func Compile(ctx context.Context, src string, out string) error {
-	debug := option.Config.BPFCompilationDebug
+	debug := option.Config.Opts.IsEnabled(option.DebugCompile)
```
As we need to keep the deprecated option until 1.11, this needs to be `debug := option.Config.BPFCompilationDebug || option.Config.Opts.IsEnabled(option.DebugCompile)`

aanm commented on pkg/datapath/loader/compile.go:
```diff
@@ -396,11 +395,13 @@ func Compile(ctx context.Context, src string, out string) error {
 
 // compileTemplate compiles a BPF program generating a template object file.
 func compileTemplate(ctx context.Context, out string, isHost bool) error {
+	debug := option.Config.Opts.IsEnabled(option.DebugCompile)
```
same here, this needs to be `debug := option.Config.BPFCompilationDebug || option.Config.Opts.IsEnabled(option.DebugCompile)`

aanm commented on pkg/datapath/loader/loader.go:
```diff
@@ -367,7 +367,8 @@ func (l *Loader) reloadDatapath(ctx context.Context, ep datapath.Endpoint, dirs
 
 func (l *Loader) compileAndLoad(ctx context.Context, ep datapath.Endpoint, dirs *directoryInfo, stats *metrics.SpanStat) error {
-	debug := option.Config.BPFCompilationDebug
+	debug := option.Config.Opts.IsEnabled(option.DebugCompile)
```
same here

qmonnet commented on daemon/cmd/daemon_main.go:
```diff
@@ -228,7 +229,8 @@ func init() {
 
-	flags.Bool(option.BPFCompileDebugName, false, "Enable debugging of the BPF compilation process")
+	flags.Bool(option.BPFCompileDebugName, false, "Enable debugging of the BPF compilation process. Deprecated in favour of --debug-verbose=datapath-compile")
+	flags.MarkDeprecated(option.BPFCompileDebugName, "This option will be removed in v1.10")
```
Do you mean v1.11?

qmonnet reviewed (changes requested):
Please double check version numbers for deprecation (see below).

It would be nice to have `--debug-verbose=datapath-compile` (and possibly the other debug groups) documented or at least mentioned somewhere in the documentation (Documentation/operations/troubleshooting.rst might be a candidate). Otherwise, I don't see how we can learn about this option without looking at the code?

qmonnet commented on Documentation/spelling_wordlist.txt:
```diff
@@ -370,6 +370,7 @@ extendable
 extensibility
 externalIPs
 fallback
+favour
```
Nit: did you try `favor` instead (US English), which would maybe spare us the addition of the word to this list?

nebril commented on test/k8sT/DatapathConfiguration.go:
```diff
@@ -610,6 +610,15 @@ var _ = Describe("K8sDatapathConfig", func() {
+			deploymentManager.DeployCilium(map[string]string{
+				"debug.verbose": "datapath-compile",
+			}, DeployCiliumOptionsAndDNS)
+			Expect(testPodConnectivityAcrossNodes(kubectl)).Should(BeTrue(), "Connectivity test between nodes failed")
```
Should we look for a specific debug output in Cilium logs in this test?

jibi replied:
Good shout :+1: 
Yes, I think I can look for `Enabling BPF compilation process debug` to confirm `--debug-verbose=datapath-compile` is actually set

borkmann commented on pkg/datapath/loader/cache.go:
```diff
@@ -65,6 +65,13 @@ var ignoredELFPrefixes = []string{
+	".debug_info",                // Debug
+	".debug_line",                // Debug
+	".debug_frame",               // Debug
+	".debug_loc",                 // Debug
```
The commit description says `as the iproute2 version shipped with Cilium is now able to parse BTF emitted by clang/llvm 10.0.0`. Could you elaborate what you mean by 'now'? It's still going through https://github.com/cilium/iproute2/commit/f823f36012fb5ab4ddfca6ed4ff56188730f281e no?

borkmann replied:
On which kernels did we test? Is there a fall-back that when a kernel doesn't accept the BTF that we sanitise it on load?

jibi replied:
> Could you elaborate what you mean by 'now'? It's still going through cilium/iproute2@f823f36 no?

I was referring to the versions we ship with the docker images:
* https://github.com/cilium/iproute2/commit/2f278887268d60fdaed3c138d779827d9c0007b4 (see [here](https://github.com/cilium/image-tools/blob/9a8bfaa6c36fb0471bb0c3484f69eefeca0bafa2/images/iproute2/checkout-iproute2.sh#L11))
* LLVM 10.0.0 (see [here](https://github.com/cilium/image-tools/blob/9a8bfaa6c36fb0471bb0c3484f69eefeca0bafa2/images/llvm/checkout-llvm.sh#L11))

> On which kernels did we test? 

So far I run a connectivity check on all CI kernels, so 4.9, 4.19, 5.4 (GKE) and net-next

> Is there a fall-back that when a kernel doesn't accept the BTF that we sanitise it on load?

There isn't as initially I assumed having debug symbols compiled in would be an opt in feature rather than something that is on by default

qmonnet reviewed (approved):
Nice, will be great to have BTF info!

aanm commented on Documentation/operations/upgrade.rst:
```diff
@@ -804,6 +804,8 @@ Deprecated options
 * ``k8s-watcher-queue-size``: This option does not have any effect since 1.8
   and is planned to be removed in 1.10.
+* ``bpf-compile-debug``: This option does not have any effect since 1.10
+  and is planned to be removed in 1.11.
```
This should be between lines 330 and 350

jibi committed:
datapath: always generate BTF debug information

Enable by default the generation of the debug symbols for the BPF
program.

The --bpf-compile-debug option does not have any effect anymore and will
be deprecated in version 1.11.

Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>

twpayne reviewed (approved):
LGTM
